sdk/Sdk.Generators/FunctionMetadataProviderGenerator.cs

```diff
@@ -0,0 +1,841 @@
+        private static void AddMethodOutputBinding(IndentedTextWriter indentedTextWriter, MethodDeclarationSyntax method, SemanticModel model, string funcName, out bool hasOutputBinding)
+        {
+            var methodSymbol = model.GetDeclaredSymbol(method);
+            var attributes = methodSymbol!.GetAttributes(); // methodSymbol is not null here because it's checked in IsMethodAFunction which is called before bindings are collected/created
+
+            AttributeData? outputBinding = null;
+            hasOutputBinding = false;
+
+            foreach (var attribute in attributes)
+            {
+                if (IsBindingAttribute(attribute))
+                {
+                    if (hasOutputBinding)
+                    {
+                        throw new FormatException($"Found multiple output attributes on method '{nameof(method)}'. Only one output binding attribute is is supported on a method.");
+                    }
+
+                    outputBinding = attribute;
+                    hasOutputBinding = true;
+                }
+            }
+
+            if (outputBinding != null)
+            {
+                WriteBindingToFile(indentedTextWriter, outputBinding, funcName, Constants.ReturnBindingName, null);
+            }
+        }
+
+        /// <summary>
+        /// This method adds the input/trigger bindings found in the parameters of the Function
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Throws when a symbol cannot be found.</exception>
+        /// <exception cref="FormatException">Throws if the attributes were used and formatted incorrectly.</exception>
+        private static void AddParameterInputAndTriggerBindings(IndentedTextWriter indentedTextWriter, MethodDeclarationSyntax method, SemanticModel model, string funcName, out bool hasHttpTrigger)
+        {
+            hasHttpTrigger = false;
+
+            foreach (ParameterSyntax parameter in method.ParameterList.Parameters)
+            {
+                // If there's no attribute, we can assume that this parameter is not a binding
+                if (parameter.AttributeLists.Count == 0)
+                {
+                    continue;
+                }
+
+                IParameterSymbol? parameterSymbol = model.GetDeclaredSymbol(parameter);
+
+                if (parameterSymbol is null)
+                {
+                    throw new InvalidOperationException($"The symbol for the parameter '{nameof(parameter)}' could not be found");
+                }
+
+                // Check to see if any of the attributes associated with this parameter is a BindingAttribute
+                foreach (var attribute in parameterSymbol.GetAttributes())
+                {
+                    if (IsBindingAttribute(attribute))
+                    {
+                        string? dataType = null;
+                        string parameterSymbolType = parameterSymbol.Type.GetFullName();
```

Comment:
nit: consider moving this closer to its usage - within the `if (dataType is null)` statement that is.